File: oasis-core/src/oasis/graphics/ogl/OglVertexBuffer.java
package oasis.graphics.ogl;

import java.nio.Buffer;

import oasis.graphics.NativeResource;
import oasis.graphics.VertexBuffer;

public class OglVertexBuffer implements NativeResource {

    public static final int TYPE = Ogl.GL_ARRAY_BUFFER; 
    
    private int[] id = new int[1]; 
    private Ogl ogl; 
    private VertexBuffer vb; 
    
    public OglVertexBuffer(Ogl ogl, VertexBuffer vb) {
        this.ogl = ogl; 
        this.vb = vb; 
    }
    
    public int getId() {
        return id[0]; 
    }
    
    /**
     * Creates buffer object if it did not exist
     * 
     * @return True if already existed
     */
    private boolean validate(Ogl ogl) {
        if (id[0] == 0) {
            // vbo does not exist, create it 
            ogl.glGenBuffers(1, id, 0); 
            return false; 
        }
        else {
            return true; 
        }
    }
    
    @Override
    public void release() {
        if (id[0] != 0) {
            // delete vbo
            ogl.glDeleteBuffers(1, id, 0); 
        }
    }

    @Override
    public void update() {
        boolean alreadyExist = validate(ogl); 
        
        int usage = OglConvert.getBufferUsage(vb.getUsage()); 
        
        ogl.glBindBuffer(TYPE, id[0]); 
        
        int bytes = vb.getSizeInBytes(); 
        Buffer data = vb.getAndFlipBuffer(); 
        
        if (alreadyExist) {
            ogl.glBufferSubData(TYPE, 0, bytes, data);
        }
        else {
            ogl.glBufferData(TYPE, bytes, data, usage);
        }
    }
    
}

File: Oasis/Graphics/OpenGL/OglVertexBuffer.cpp
#include "Oasis/Graphics/OpenGL/OglVertexBuffer.h"

#include <string.h>

#include <iostream>

using namespace std;

namespace Oasis
{

OglVertexBuffer::OglVertexBuffer(int elements, const VertexFormat& format, BufferUsage usage)
    : VertexBuffer(elements, format, usage)
    , m_id(0)
{
    glGenBuffers(1, &m_id);
    glBindBuffer(GL_ARRAY_BUFFER, m_id);
    glBufferData(GL_ARRAY_BUFFER, elements * format.GetSize() * sizeof (float), NULL, GL_DYNAMIC_DRAW);
}

OglVertexBuffer::~OglVertexBuffer() { ReleaseGpuData(); }

void OglVertexBuffer::ReleaseGpuData()
{
    if (m_id)
    {
        glDeleteBuffers(1, &m_id);
        m_id = 0;
    }
}

void OglVertexBuffer::UploadGpuData(int bytes, const void* data)
{
    glBindBuffer(GL_ARRAY_BUFFER, m_id);
    glBufferData(GL_ARRAY_BUFFER, bytes, data, GL_DYNAMIC_DRAW);
}

void OglVertexBuffer::UploadGpuSubData(int offset, int bytes, const void* data)
{
    glBindBuffer(GL_ARRAY_BUFFER, m_id);
    glBufferSubData(GL_ARRAY_BUFFER, offset, bytes, data);
}

}
